crypto/stack/stack.c

```diff
@@ -107,43 +112,99 @@ OPENSSL_STACK *OPENSSL_sk_new(OPENSSL_sk_compfunc c)
 
     if ((ret = OPENSSL_zalloc(sizeof(*ret))) == NULL)
         goto err;
-    if ((ret->data = OPENSSL_zalloc(sizeof(*ret->data) * MIN_NODES)) == NULL)
+    if ((ret->data = OPENSSL_zalloc(sizeof(*ret->data) * min_nodes)) == NULL)
         goto err;
     ret->comp = c;
-    ret->num_alloc = MIN_NODES;
+    ret->num_alloc = min_nodes;
     return (ret);
 
  err:
     OPENSSL_free(ret);
     return (NULL);
 }
 
-int OPENSSL_sk_insert(OPENSSL_STACK *st, const void *data, int loc)
+/*
+ * Calculate the array growth based on the target size.
+ *
+ * The growth faction is a rational number and is defined by a numerator
+ * and a denominator.  According to Andrew Koenig in his paper "Why Are
+ * Vectors Efficient?" from JOOP 11(5) 1998, this factor should be less
+ * than the golden ratio (1.618...).
+ *
+ * We use 3/2 = 1.5 for simplicty of calculation and overflow checking.
+ * Another option 8/5 = 1.6 allows for slightly faster growth, although safe
+ * computation is more difficult.
+ *
+ * The limit to avoid overflow is spot on.  The modulo three correction term
+ * ensures that the limit is the largest number than can be expanded by the
+ * growth factor without exceeding the hard limit.
+ */
+static ossl_inline int compute_growth(int target, int current)
 {
-    if (st == NULL || st->num < 0 || st->num == INT_MAX) {
-        return 0;
+    const int limit = (max_nodes / 3) * 2 + max_nodes % 3;
```

Comment:
Now when max_nodes is "variable" `max_nodes % 3` term appears somewhat controversial. For known values of 0x7fffffff and 0xffffffff/4 it works, but not in general case. "Works" means that `limit + limit/2` doesn't actually exceed the max_nodes. Or in other words *in general case* you can find a max_nodes value for which `limit + limit/2` would exceed max_nodes. So that preceding commentary is actually misleading, because it gives an impression of general case. On the other hand we are not performing `limit + limit/2` operation, but `limit-1 + (limit-1)/2`... So I suppose suggestion is to either omit the term, or clarify commentary by stating that we are scaling up limit-1 [because it's `current < limit` and not `current <= limit`].

Reply:
Ouch!  2^n mod 3 is always 0 or 1.  The code as written overestimates by 1 when `max_nodes` mod 3 = 2.  Good catch.

Reply:
> 2^n mod 3 is always 0 or 1

??? 2^3 % 3 is 2.

Reply:
2^n-1 was the intention.